File: Algorithms/Medium/HashTables/890. Find and Replace Pattern/python/solution.py
from typing import Pattern


class Solution(object):
    def findAndReplacePattern(self, words, pattern):
        """
        :type words: List[str]
        :type pattern: str
        :rtype: List[str]
        """
        result = []
        
        for word in words: 
            if len(word) != len(pattern):
                 continue
            
            if self.patternMatch(word, pattern):
                result.append(word)

        return result

    def patternMatch(self, word, pattern):
        """
        """
        dict = {}
        mapped = set()        

        i = 0
        while i < len(word):
            p = pattern[i]
            w = word[i]

            if p in dict:
                if (w != dict[p]):
                     return False
            else:
                if (w in mapped):
                     return False

                mapped.add(w)
                dict[p] = w                

            i += 1

        return True

words = ["abc","deq","mee","aqq","dkd","ccc"]
pattern = "abb"

s = Solution()
list = s.findAndReplacePattern(words, pattern)
print(list)

File: Algorithms/Medium/HashTables/890. Find and Replace Pattern/Solution.cs
public class Solution 
{
    public IList<string> FindAndReplacePattern(string[] words, string pattern) 
    {
        var result = new List<string>();

        foreach (var word in words)
        {
            if (word.Length != pattern.Length) continue;

            if (PatternMatch(word, pattern))
            {
                result.Add(word);
            }
        }

        return result;
    }

    private bool PatternMatch(string word, string pattern)
    {
        var dict = new Dictionary<char, char>();
        var mapped = new HashSet<char>();

        for (int i = 0; i < word.Length; i++)
        {
            char p = pattern[i];
            char w = word[i];

            if (dict.ContainsKey(p))
            {
                if (w != dict[p]) return false;
            }
            else
            {
                if (mapped.Contains(w)) return false;

                mapped.Add(w);
                dict.Add(p, w);
            }
        }

        return true;
    }
}